rust-src/wallet_library/src/statement.rs

```diff
@@ -0,0 +1,274 @@
+
+fn is_iso8601(date: &str) -> bool {
+    println!("{}", date);
+    chrono::NaiveDate::parse_from_str(date, "%Y%m%d").is_ok()
+}
+
+fn is_iso3166_alpha_2(code: &str) -> bool { rust_iso3166::from_alpha2(code).is_some() }
+
+fn is_iso3166_2(code: &str) -> bool { rust_iso3166::iso3166_2::from_code(code).is_some() }
+
+pub trait WalletCheck {
+    fn satisfies_wallet_restrictions(&self) -> bool;
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> WalletCheck for Request<C, AttributeType> {
+    fn satisfies_wallet_restrictions(&self) -> bool {
+        self.credential_statements
+            .iter()
+            .all(|s| s.satisfies_wallet_restrictions())
+    }
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> WalletCheck
+    for CredentialStatement<C, AttributeType>
+{
+    fn satisfies_wallet_restrictions(&self) -> bool {
+        match self {
+            CredentialStatement::Account {
+                statement,
+                network: _,
+                cred_id: _,
+            } => {
+                if statement.is_empty() {
+                    return false;
+                }
+                let mut used_tags = Vec::<AttributeTag>::new();
+                for atomic_statement in statement {
+                    if used_tags.contains(&atomic_statement.attribute()) {
+                        return false;
+                    }
+                    used_tags.push(atomic_statement.attribute());
+
+                    if !atomic_statement.satisfies_wallet_restrictions() {
+                        return false;
+                    }
+                }
+            }
+            CredentialStatement::Web3Id {
+                ty: _,
+                network: _,
+                contract: _,
+                credential: _,
+                statement: _,
+            } => todo!(),
+        }
+        true
+    }
+}
+
+// Note that this is an implementation only for the account statement, the tag
+// is AttributeTag.
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> WalletCheck
+    for AtomicStatement<C, AttributeTag, AttributeType>
+{
+    fn satisfies_wallet_restrictions(&self) -> bool {
+        let check_attribute_value = |value: &AttributeType| match self.attribute().0 {
+            // countryOfResidence | nationality
+            4 | 5 => is_iso3166_alpha_2(&value.to_string()),
+            // idDocIssuer
+            8 => is_iso3166_alpha_2(&value.to_string()) || is_iso3166_2(&value.to_string()),
+            // dob, idDocIssuedAt, idDocExpiresAt
+            3 | 9 | 10 => is_iso8601(&value.to_string()),
+            _ => true,
+        };
+
+        match self {
+            AtomicStatement::RevealAttribute { statement: _ } => true,
+            AtomicStatement::AttributeInRange { statement } => {
+                check_attribute_value(&statement.lower)
+                    && check_attribute_value(&statement.upper)
+                    && statement.lower < statement.upper
```

Comment:
WHy not <=?

Reply:
Hmm I thought that lower = upper doesn't make sense, when upper bound is exclusive? :thinking: 
We don't allow it in javascript. :upside_down_face:

Reply:
Yeah you are right, upper bound is exclusive.